tutorials/intermediate/imshow_extent.py

```diff
@@ -0,0 +1,261 @@
+                       'xycoords': 'axes fraction',
+                       'xy': (1, .5)}
+        if extent is None:
+            ax.annotate('None', **text_kwargs)
+            ax.set_title('extent=')
+        else:
+            left, right, bottom, top = extent
+            text = ('left: {left:0.1f}\nright: {right:0.1f}\n' +
+                    'bottom: {bottom:0.1f}\ntop: {top:0.1f}\n').format(
+                        left=left, right=right, bottom=bottom, top=top)
+
+            ax.annotate(text, **text_kwargs)
+        ax.axis('off')
+    return columns
+
+
+
+###############################################################################
+#
+# Default extent
+# --------------
+#
+# First, let's have a look at the default `extent=None`
+
+generate_imshow_demo_grid(extents=[None])
+
+###############################################################################
+#
+# Generally, for an array of shape (M, N), the first index runs along the
+# vertical, the second index runs along the horizontal.
+# The pixel centers are at integer positions ranging from 0 to `N' = N - 1``
+# horizontally and from 0 to ``M' = M - 1`` vertically.
+# *origin* determines how to the data is filled in the bounding box.
+#
+# For ``origin='lower'``:
+#
+#    - [0, 0] is at (left, bottom)
+#    - [M', 0] is at (left, top)
+#    - [0, N'] is at (right, bottom)
+#    - [M', N'] is at (right, top)
+#
+# ``origin='upper'`` reverses the vertical axes direction and filling:
+#
+#    - [0, 0] is at (left, top)
+#    - [M', 0] is at (left, bottom)
+#    - [0, N'] is at (right, top)
+#    - [M', N'] is at (right, bottom)
+#
+# In summary, the position of the [0, 0] index as well as the extent are
+# influenced by *origin*:
+#
+# ======  ===============  ==========================================
+# origin  [0, 0] position  extent
+# ======  ===============  ==========================================
+# upper   top left         ``(-0.5, numcols-0.5, numrows-0.5, -0.5)``
+# lower   bottom left      ``(-0.5, numcols-0.5, -0.5, numrows-0.5)``
+# ======  ===============  ==========================================
+#
+# The default value of *origin* is set by :rc:`image.origin` which defaults
+# to ``'upper'`` to match the matrix indexing conventions in math and
+# computer graphics image indexing conventions.
+#
+#
+# Explicit extent
+# ---------------
+#
+# By setting *extent* we pick a bounding box in dataspace that the image will
+# fill. Then, the underlying data is interpolated/resampled to fill that box.
+#
+# If the axes is set to autoscale, then view limits of the axes are set
+# to match the *extent* which ensures that the coordinate set by
+# ``(left, bottom)`` is at the bottom left of the axes!  However, this
+# may invert the axis so they do not increase in the 'natural' direction.
+#
+
+extents = [(-0.5, 6.5, -0.5, 5.5),
+           (-0.5, 6.5, 5.5, -0.5),
+           (6.5, -0.5, -0.5, 5.5),
+           (6.5, -0.5, 5.5, -0.5)]
+
```

Comment:
This set of examples doesn't really help me because the "extents" is still the same as the extents I'd expect from the underlying data array size.  Wouldn't it be more illustrative to go from 0-15 or something?  The phrase "...a bounding box in data space" is also mysterious.  I think of the matrix as being the data and a bounding box on the data means a clip to me.  But what you mean is that the extents are the x and y co-ordinates over which the matrix is placed.  

The docs say "The location, in data-coordinates, of the lower-left and upper-right corners" ; I prefer "data-coordinates"

Reply:
The point is to show how inverting the left vs right or top vs bottom changes the image around.  Using different limits would make that a harder apples-to-apples comparison.

Reply:
yes But it didn’t help me understand what extents actually is on the first place

I’ll admit this flipping the axes seems like the wrong behaviour.  I’d have expected the matrix to flip.  But I doubt I’ll get that changed here.  ;-)